components/org.wso2.carbon.identity.password.expiry/src/main/java/org/wso2/carbon/identity/password/expiry/util/PasswordPolicyUtils.java

```diff
@@ -292,7 +293,135 @@ private static boolean isPasswordExpiredUnderDefaultPolicy(String tenantDomain,
+            if (CollectionUtils.isEmpty(passwordExpiryRules)) {
+                if (isSkipIfNoApplicableRulesEnabled) return Optional.empty();
+                // If lastPasswordUpdatedTime is blank, set expiry time to now.
+                if (isLastPasswordUpdatedTimeBlank) {
+                    return Optional.of(System.currentTimeMillis());
+                }
+                return Optional.of(
+                        lastPasswordUpdatedTimeInMillis + getDaysTimeInMillis(defaultPasswordExpiryInDays));
+            }
+
+            Map<PasswordExpiryRuleAttributeEnum, Set<String>> userAttributes =
+                    new EnumMap<>(PasswordExpiryRuleAttributeEnum.class);
+
+            List<PasswordExpiryRule> filteredRules =
+                    filterApplicableExpiryRules(passwordExpiryRules, isSkipIfNoApplicableRulesEnabled);
+            for (PasswordExpiryRule rule : filteredRules) {
+                if (isRuleApplicable(rule, userAttributes, tenantDomain, userId, userStoreManager)) {
+                    // Skip the rule if the operator is not equals.
+                    if (PasswordExpiryRuleOperatorEnum.NE.equals(rule.getOperator())) {
+                        return Optional.empty();
+                    }
+                    if (isLastPasswordUpdatedTimeBlank) {
+                        return Optional.of(System.currentTimeMillis());
+                    }
+                    int expiryDays =
+                            rule.getExpiryDays() > 0 ? rule.getExpiryDays() : getPasswordExpiryInDays(tenantDomain);
+                    return Optional.of(lastPasswordUpdatedTimeInMillis + getDaysTimeInMillis(expiryDays));
+                }
+            }
+
+            if (isSkipIfNoApplicableRulesEnabled) return Optional.empty();
+            if (isLastPasswordUpdatedTimeBlank) {
+                return Optional.of(System.currentTimeMillis());
+            }
+            return Optional.of(
+                    lastPasswordUpdatedTimeInMillis + getDaysTimeInMillis(defaultPasswordExpiryInDays));
+        } catch (UserStoreException e) {
+            throw new PostAuthenticationFailedException(PasswordPolicyConstants.ErrorMessages.
+                    ERROR_WHILE_GETTING_USER_STORE_DOMAIN.getCode(),
+                    PasswordPolicyConstants.ErrorMessages.ERROR_WHILE_GETTING_USER_STORE_DOMAIN.getMessage());
+        }
+    }
+
+    private static List<PasswordExpiryRule> filterApplicableExpiryRules(List<PasswordExpiryRule> passwordExpiryRules,
+                                                                        boolean skipIfNoApplicableRules) {
+
+        // If the default behavior is to skip the password expiry, rules with skip logic are not required.
+        return passwordExpiryRules.stream().filter(
+                rule -> !skipIfNoApplicableRules || !PasswordExpiryRuleOperatorEnum.NE.equals(rule.getOperator()))
```

Comment:
`!skipIfNoApplicableRules` can we get this logic out of the filter condition, since it does not depend on each rule?

Reply:
Fixed